geoh5py/ui_json/enforcers.py

```diff
@@ -0,0 +1,350 @@
+        """
+        Create enforcers pool from name/validation dictionary.
+
+        :param validations: Encodes validations as enforcer type and
+            validation key value pairs.
+        :param protected: Excludes listed enforcer types from updating
+
+        """
+        protected = protected or []
+        enforcers = [k for k in self.enforcers if k.enforcer_type in protected]
+        for name, validation in validations.items():
+            if name not in protected:
+                enforcers.append(self._recruit_enforcer(name, validation))
+
+        self.enforcers = enforcers
+
+    @property
+    def validations(self) -> dict[str, Any]:
+        """Returns an enforcer type / validation dictionary from pool."""
+        return {k.enforcer_type: k.validations for k in self.enforcers}
+
+    def _recruit_enforcer(self, enforcer_type: str, validation: Any) -> Enforcer:
+        """
+        Create enforcer from enforcer type and validation.
+
+        :param enforcer_type: Type of enforcer to create.
+        :param validation: Enforcer validation.
+        """
+
+        if enforcer_type not in self.enforcer_types:
+            raise ValueError(f"Invalid enforcer type: {validation['type']}")
+
+        if enforcer_type == "type":
+            enforcer = TypeEnforcer(validation)
+        if enforcer_type == "value":
+            enforcer = ValueEnforcer(validation)  # type: ignore
+        if enforcer_type == "uuid":
+            enforcer = UUIDEnforcer(validation)  # type: ignore
+        if enforcer_type == "required":
+            enforcer = RequiredEnforcer(validation)  # type: ignore
+        if enforcer_type == "required_uijson_parameters":
+            enforcer = RequiredUIJsonParameterEnforcer(validation)  # type: ignore
+        if enforcer_type == "required_form_members":
+            enforcer = RequiredFormMemberEnforcer(validation)  # type: ignore
+        if enforcer_type == "required_workspace_objects":
+            enforcer = RequiredWorkspaceObjectEnforcer(validation)  # type: ignore
+        if enforcer_type == "required_object_data":
+            enforcer = RequiredObjectDataEnforcer(validation)  # type: ignore
+
+        return enforcer
+
+    def validate(self, value: Any):
+        """Validate value against all enforcers."""
+
+        for enforcer in self.enforcers:
+            self._capture_error(enforcer, value)
+
+        self._raise_errors()
+
+    def _capture_error(self, enforcer: Enforcer, value: Any):
+        """Catch and store 'BaseValidationError's for aggregation."""
+        try:
+            enforcer.enforce(self.name, value)
+        except BaseValidationError as err:
+            self._errors.append(err)
+
+    def _raise_errors(self):
+        """Raise errors if any exist, aggregate if more than one."""
+        if self._errors:
+            if len(self._errors) > 1:
+                raise AggregateValidationError(self.name, self._errors)

```

Comment:
Is there an efficiency reason to not always just go through the AggregateValidationError when just one error? Would make the code simpler.

Reply:
It's just more readable when the raise occurs for the user.  I think I prefer complexity in the code rather than in an error message that we needs to parse just to find out that a TypeValidationError was thrown on the value of parameter x. 

The plan is to extract a class here that we can re-use when we implement the uijson update values.   since we'll need to aggregate the errors there too.  That will restore the single responsability principle for this method!